Fix PMMP Bedrock Servers Query

AlphaBaqpla opened the pull request:
I double-checked everything and fixed the error due to which the validation test did not pass

kartashovio commented:
This fixes the issue #433

PerchunPak commented:
Yes, tests do not pass because of (maybe) invalid `poetry.lock`.

ItsDrike commented:
Blocked by #440, which fixes the issues with CI checks

AlphaBaqpla commented:
The changes have been tested for 12 hours and there are no problems with them

AlphaBaqpla commented:
hmm, 1 test failed, 2 cancelled. why did it pass all the tests in my repository

ItsDrike reviewed (approved):
Seems alright to me. Thanks @AlphaBaqpla!

PerchunPak commented:
Could someone also describe what exactly this change do, why it works and what did it change.

P.S. My idea is kind of change in protocol.

AlphaBaqpla commented:
Right now, the wrong magic packet is being used when sending a request. There was no problem with GeyserMC and Bedrock Dedicated Server, but when working with PMMP it complained about the wrong packet and dropped the connection. I corrected it to the correct one and everything began to work correctly.

PerchunPak commented:
I mean, why did it work? What bytes/information was skipped/incorrect?

AlphaBaqpla commented:
https://github.com/facebookarchive/RakNet/blob/1a169895a900c9fc4841c556e16514182b75faf8/Source/RakPeer.cpp#L135
Look at this, I just modified the magic data so that it was like this.

AlphaBaqpla commented:
and pmmp itself uses this to get the status of other servers

PerchunPak commented:
> https://github.com/facebookarchive/RakNet/blob/1a169895a900c9fc4841c556e16514182b75faf8/Source/RakPeer.cpp#L135 Look at this, I just modified the magic data so that it was like this.

I know that you just copy pasted code from some library. That is the problem. No one (probably) doesn't understand what was changed and why it works now.

> and pmmp itself uses this to get the status of other servers

Could you give a link please?

AlphaBaqpla commented:
https://github.com/jasonwynn10/libpmquery/blob/master/src/libpmquery/PMQuery.php

PerchunPak commented:
It is just a library for PMMP plugins with a questionable and small Git history. So it is definitely not a reliable source.

Would be much better to reverse-engineer the parsing system of PMMP itself.

AlphaBaqpla commented:
> Would be much better to reverse-engineer the parsing system of PMMP itself.
https://wiki.vg/Raknet_Protocol
https://github.com/pmmp/PocketMine-MP/blob/stable/tools/ping-server.php

PerchunPak commented:
If diff code before and your one, the magic part (third one) is the same as was
![image](https://user-images.githubusercontent.com/68118654/208946949-6dc1eb30-1c80-4594-a7d7-21af2dcd51e0.png)

On image, there are four parts, as should be, if reference to protocol (numbers are their position).
[![image](https://user-images.githubusercontent.com/68118654/208947369-1bbfefd4-4337-4dab-b8bf-962d5626c25a.png)](https://wiki.vg/Raknet_Protocol#Unconnected_Ping)

So you actually modified the second and the the fourth part.
![image](https://user-images.githubusercontent.com/68118654/208947646-707c9737-1c4e-4ded-b413-90b911316dfc.png)

Actual problem here, that our old solution lacks last 8 bytes - Client GUID.

ItsDrike commented:
To address some of the confusion around this (though I'll admit I don't have a huge experience with it, but this is at least what I've found, in collaboration with @PerchunPak). 

The `request_status_data` variable is holding data of the entire packet that we will be sending. Specifically, it is the `Unconnected Ping` packet. The format of this packet (according to https://wiki.vg/Raknet_Protocol#Unconnected_Ping) is the following:
![image](https://user-images.githubusercontent.com/20902250/208947604-788776e9-6c1c-49f6-ba26-4baa9aa07850.png)
With the `MAGIC` being a constant value, of exactly these hex bytes: `00ffff00fefefefefdfdfdfd12345678`.

So, these are the data this PR proposes we use:
![image](https://user-images.githubusercontent.com/20902250/208947765-0f0ceedd-3f13-47a8-9fcd-38a7ca271e16.png)

Whereas previously, we were using this:
![image](https://user-images.githubusercontent.com/20902250/208948213-969393df-6067-4cf6-91db-e781fdd80669.png)

The time data probably didn't need to be changed, and could've stayed as 0, what's however interesting is that our original approach was clearly completely missing the last 8 bytes of data, being the client GUID.  I don't know why that was the case, and I don't even know how come it was even working without them, if it's clearly a part of the packet specification, but that is what this PR actually adds.

I don't have much experience with bedrock, and I honestly don't really know what the client GUID even is, however as it is a part of the packet format, and since from my limited testing, it seems to be working fine with other servers bedrock servers, I think this can be merged.

PerchunPak commented on mcstatus/bedrock_status.py:
```diff
@@ -11,7 +11,9 @@
 
 class BedrockServerStatus:
-    request_status_data = b"\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\xfe\xfe\xfe\xfe\xfd\xfd\xfd\xfd\x124Vx"
+    request_status_data = (
+        b"\x01\x00\x00\x00\x00\x00\xda\\\xb1\x00\xff\xff\x00\xfe\xfe\xfe\xfe\xfd\xfd\xfd\xfd\x124Vxu4] \xf7B\xb9\xf4"
+    )
```
If count my previous comment, we should use
```suggestion
    request_status_data = bytes.fromhex(
        # see https://wiki.vg/Raknet_Protocol#Unconnected_Ping
        "01" + "000000000000000000" + "ffff00fefefefefdfdfdfd12345678" + "00000000000000"
    )
```

ItsDrike replied:
I do agree that this would be a much cleaner approach, that actually demonstrates what's going on if someone would take a look at the linked packet specification.

AlphaBaqpla replied:
Previously, there was no client GUID, because of this the package was not complete and PMMP blocked it, unlike Bedrock Dedicated Server. The GUID can be set as random 8 bytes and it is not required to change it with each request

PerchunPak replied:
> Previously, there was no client GUID, because of this the package was not complete and PMMP blocked it, unlike Bedrock Dedicated Server. The GUID can be set as random 8 bytes and it is not required to change it with each request

So can we set it to zeros? If so, let's follow the previous "style" (the time field contained only zeros).

P.S. The proposed change doesn't work, use `0000000000000000` in the end. I had 7 bytes, but the protocol requires 8.

AlphaBaqpla replied:
> So can we set it to zeros? If so, let's follow the previous "style" (the time field contained only zeros).
> 
> P.S. The proposed change doesn't work, use `0000000000000000` in the end. I had 7 bytes, but the protocol requires 8.

As the GUID section, need to use random 8 bytes (not zeros) - "a20871356d207e61" for example.
As the time section, can used only zeros

AlphaBaqpla replied:
Hmm, request from https://bedrockinfo.com/
![image](https://user-images.githubusercontent.com/82477148/209293259-09ce95df-d321-487b-a578-e5f55708b014.png)
only zeros as GUID and time

PerchunPak replied:
> As the GUID section, need to use random 8 bytes (not zeros) - "a20871356d207e61" for example. (this will work only with zeros, but it's not quite right)
> As the time section, can used only zeros

We have #448 for refactor of this, for now, let's use zeros.

AlphaBaqpla replied:
> We have #448 for refactor of this, for now, let's use zeros.
yes

PerchunPak commented:
> I honestly don't really know what the client GUID even is

I investigated [CloudburstMC/Network](https://github.com/CloudburstMC/Network) and found that they just [generate random value as a client GUID](https://github.com/CloudburstMC/Network/blob/ba2eba88e0ab0f14759a8edcf43965d3a3387bcc/raknet/src/main/java/com/nukkitx/network/raknet/RakNet.java#L27).

PerchunPak commented on mcstatus/bedrock_status.py:
```diff
@@ -11,7 +11,10 @@
-    request_status_data = b"\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\xfe\xfe\xfe\xfe\xfd\xfd\xfd\xfd\x124Vx"
+    request_status_data = bytes.fromhex(
+        # see https://wiki.vg/Raknet_Protocol#Unconnected_Ping
+        "01" + "000000000000000000" + "ffff00fefefefefdfdfdfd12345678" + "a20871356d207e61"
```
Black change terrible here, let's skip it.
```suggestion
        "01" + "000000000000000000" + "ffff00fefefefefdfdfdfd12345678" + "a20871356d207e61"  # fmt: skip
```

ItsDrike commented on mcstatus/bedrock_status.py:
```diff
@@ -11,7 +11,10 @@
-    request_status_data = b"\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\xfe\xfe\xfe\xfe\xfd\xfd\xfd\xfd\x124Vx"
+    request_status_data = bytes.fromhex(
+        # see https://wiki.vg/Raknet_Protocol#Unconnected_Ping
+        "01" + "000000000000000000" + "ffff00fefefefefdfdfdfd12345678" + "a20871356d207e61"  # fmt: skip
```
```suggestion
        "01" + "000000000000000000" + "ffff00fefefefefdfdfdfd12345678" + "a20871356d207e61"  # fmt: skip
```

The PR looks ok, however we could replace the random data with all 0es instead. I know it's supposed to be something random, but all zeroes are one of the things full randomized data can give you. 

Though hard-coding some gibberish like this doesn't help make it clear what this is, and it actually makes it appear as if this data was somehow deliberately chosen, when it's really just random. Let's use 0es for now.

AlphaBaqpla replied:
yes, it is quite possible to replace, this method is used and works
